beacon-chain/core/blocks/block_operations.go

```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
+//   Verify that len(block.body.deposits) <= MAX_DEPOSITS.
+//   For each deposit in block.body.deposits:
+//     Let serialized_deposit_data be the serialized form of deposit.deposit_data.
+//     It should be the DepositInput followed by 8 bytes for deposit_data.value
+//     and 8 bytes for deposit_data.timestamp. That is, it should match
+//     deposit_data in the Ethereum 1.0 deposit contract of which the hash
+//     was placed into the Merkle tree.
+//
+//     Verify deposit merkle_branch, setting leaf=serialized_deposit_data,
+//     depth=DEPOSIT_CONTRACT_TREE_DEPTH and root=state.processed_pow_receipt_root:
+//     Verify that state.slot - (deposit.deposit_data.timestamp -
+//     state.genesis_time)  SLOT_DURATION < ZERO_BALANCE_VALIDATOR_TTL.
+//
+//     Run the following:
+//     process_deposit(
+//       state=state,
+//       pubkey=deposit.deposit_data.deposit_input.pubkey,
+//       deposit=deposit.deposit_data.value,
+//       proof_of_possession=deposit.deposit_data.deposit_input.proof_of_possession,
+//       withdrawal_credentials=deposit.deposit_data.deposit_input.withdrawal_credentials,
+//       randao_commitment=deposit.deposit_data.deposit_input.randao_commitment,
+//       poc_commitment=deposit.deposit_data.deposit_input.poc_commitment,
+//     )
+func ProcessBlockValidatorDeposits(
+	beaconState *pb.BeaconState,
+	block *pb.BeaconBlock,
+) (*pb.BeaconState, error) {
+	deposits := block.GetBody().GetDeposits()
+	if uint64(len(deposits)) > params.BeaconConfig().MaxDeposits {
+		return nil, fmt.Errorf(
+			"number of deposits (%d) exceeds allowed threshold of %d",
+			len(deposits),
+			params.BeaconConfig().MaxDeposits,
+		)
+	}
+	var err error
+	var depositInput *pb.DepositInput
+	for idx, deposit := range deposits {
+		depositData := deposit.GetDepositData()
+		depositInput, err = DecodeDepositInput(depositData)
+		if err != nil {
+			return nil, fmt.Errorf("could not decode deposit input: %v", err)
+		}
+		if err = verifyDeposit(beaconState, deposit); err != nil {
+			return nil, fmt.Errorf("could not verify deposit #%d: %v", idx, err)
+		}
+		// depositData consists of depositInput []byte + depositValue [8]byte +
+		// depositTimestamp [8]byte.
+		depositValue := depositData[len(depositData)-16 : len(depositData)-8]
+		// We then mutate the beacon state with the verified validator deposit.
+		beaconState, _, err = v.ProcessDeposit(
+			beaconState,
+			depositInput.GetPubkey(),
+			binary.BigEndian.Uint64(depositValue),
+			depositInput.GetProofOfPossession(),
+			depositInput.GetWithdrawalCredentialsHash32(),
+			depositInput.GetRandaoCommitmentHash32(),
+			depositInput.GetPocCommitment(),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could process deposit into beacon state: %v", err)
+		}
+	}
+	return beaconState, nil
+}
+
+// DecodeDepositInput unmarshales a depositData byte slice into
+// a proto *pb.DepositInput by using the Simple Serialize (SSZ)
```

Comment:
the depositData we get from lighthouse won't be in serialized proto format, but we can work interoperability with other client implementations later